File: qt-app/src/services/optimization/Optimizer.cpp
#include "Optimizer.h"

#include <algorithm>

#include "../../models/Asset.h"
#include "../../utils/Math.h"
#include "../data/DataStore.h"

Optimizer::Optimizer(DataStore* dataStore)
    : m_dataStore(dataStore)
{
}

QVector<Holding> Optimizer::optimize(double targetYield, int maxAssets) const
{
    if (m_dataStore == nullptr || maxAssets <= 0) {
        return {};
    }

    QVector<Asset> universe = m_dataStore->loadAssets();
    universe.erase(std::remove_if(universe.begin(), universe.end(), [](const Asset& asset) {
        return asset.dividendYield <= 0.0;
    }), universe.end());

    std::sort(universe.begin(), universe.end(), [](const Asset& left, const Asset& right) {
        if (!qFuzzyCompare(left.beta + 1.0, right.beta + 1.0)) {
            return left.beta < right.beta;
        }
        return left.dividendYield > right.dividendYield;
    });

    if (universe.isEmpty()) {
        return {};
    }

    QVector<Holding> holdings;
    QVector<Asset> qualifying;
    for (const auto& asset : universe) {
        if (asset.dividendYield + 1e-6 >= targetYield) {
            qualifying.append(asset);
        }
    }

    const QVector<Asset>& selection = qualifying.isEmpty() ? universe : qualifying;
    const int selectionCount = qMin(maxAssets, selection.size());
    for (int i = 0; i < selectionCount; ++i) {
        holdings.append({selection[i].ticker, 0.0, selection[i].dividendYield, selection[i].beta});
    }

    if (holdings.isEmpty()) {
        return {};
    }

    if (!qualifying.isEmpty()) {
        holdings[0].weight = 1.0;
        holdings.resize(1);
        return holdings;
    }

    const double equalWeight = 1.0 / static_cast<double>(holdings.size());
    for (auto& holding : holdings) {
        holding.weight = equalWeight;
    }

    double currentYield = portfolioYield(holdings);
    for (int iteration = 0; iteration < 250 && currentYield + 1e-6 < targetYield; ++iteration) {
        int lowYieldIndex = 0;
        int highYieldIndex = 0;
        for (int i = 1; i < holdings.size(); ++i) {
            if (holdings[i].yield < holdings[lowYieldIndex].yield ||
                (qFuzzyCompare(holdings[i].yield + 1.0, holdings[lowYieldIndex].yield + 1.0) && holdings[i].beta > holdings[lowYieldIndex].beta)) {
                lowYieldIndex = i;
            }
            if (holdings[i].yield > holdings[highYieldIndex].yield ||
                (qFuzzyCompare(holdings[i].yield + 1.0, holdings[highYieldIndex].yield + 1.0) && holdings[i].beta < holdings[highYieldIndex].beta)) {
                highYieldIndex = i;
            }
        }

        if (lowYieldIndex == highYieldIndex) {
            break;
        }

        const double shift = qMin(0.02, holdings[lowYieldIndex].weight);
        if (shift <= 0.0) {
            break;
        }

        holdings[lowYieldIndex].weight -= shift;
        holdings[highYieldIndex].weight += shift;
        currentYield = portfolioYield(holdings);
    }

    double totalWeight = 0.0;
    for (const auto& holding : holdings) {
        totalWeight += holding.weight;
    }

    if (totalWeight > 0.0) {
        for (auto& holding : holdings) {
            holding.weight /= totalWeight;
        }
    }

    holdings.erase(std::remove_if(holdings.begin(), holdings.end(), [](const Holding& holding) {
        return holding.weight <= 1e-6;
    }), holdings.end());

    std::sort(holdings.begin(), holdings.end(), [](const Holding& left, const Holding& right) {
        return left.weight > right.weight;
    });

    return holdings;
}

File: src/services/optimizer.py
"""Portfolio optimizer: yield-targeted allocation with beta minimization."""

from typing import List

from ..models import Asset, Holding


def portfolio_yield(holdings: List[Holding]) -> float:
    total = sum(h.weight * h.dividend_yield for h in holdings)
    return total


def portfolio_beta(holdings: List[Holding]) -> float:
    total = sum(h.weight * h.beta for h in holdings)
    return total


def optimize(assets: List[Asset], target_yield: float, max_assets: int = 5) -> List[Holding]:
    """Generate optimal portfolio holdings for given target yield.
    
    Strategy: select lowest-beta assets that meet yield target,
    then iteratively shift weight toward higher-yielding assets.
    """
    if not assets or max_assets <= 0:
        return []

    # Filter to dividend-paying assets only
    universe = [a for a in assets if a.dividend_yield > 0]
    if not universe:
        return []

    # Sort by beta ascending, dividend_yield descending as tiebreaker
    universe.sort(key=lambda a: (a.beta, -a.dividend_yield))

    # Find qualifying assets (yield >= target)
    qualifying = [a for a in universe if a.dividend_yield >= target_yield - 1e-6]

    if qualifying:
        # Single best asset: lowest beta among qualifying
        best = min(qualifying, key=lambda a: a.beta)
        return [Holding(best.ticker, 1.0, best.dividend_yield, best.beta)]

    # Take top N lowest-beta assets and optimize weights
    selection = universe[:max_assets]
    holdings = [Holding(a.ticker, 1.0 / len(selection), a.dividend_yield, a.beta) for a in selection]

    # Iteratively shift weight from low-yield to high-yield holdings
    for _ in range(250):
        current_yield = portfolio_yield(holdings)
        if current_yield + 1e-6 >= target_yield:
            break

        low_idx = min(range(len(holdings)), key=lambda i: holdings[i].dividend_yield)
        high_idx = max(range(len(holdings)), key=lambda i: holdings[i].dividend_yield)

        if low_idx == high_idx:
            break

        shift = min(0.02, holdings[low_idx].weight)
        if shift <= 0:
            break

        holdings[low_idx].weight -= shift
        holdings[high_idx].weight += shift

    # Normalize and clean up
    total_weight = sum(h.weight for h in holdings)
    if total_weight > 0:
        for h in holdings:
            h.weight /= total_weight

    holdings = [h for h in holdings if h.weight > 1e-6]
    holdings.sort(key=lambda h: -h.weight)
    return holdings
